opentelemetry/proto/collector/dynamicconfig/v1/dynamicconfig_service.proto

```diff
@@ -0,0 +1,98 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package opentelemetry.proto.collector.dynamicconfig.v1;
+
+import "opentelemetry/proto/resource/v1/resource.proto";
+
+option java_multiple_files = true;
+option java_package = "io.opentelemetry.proto.collector.dynamicconfig.v1";
+option java_outer_classname = "DynamicConfigServiceProto";
+option go_package = "github.com/open-telemetry/opentelemetry-proto/gen/go/collector/dynamicconfig/v1";
+
+// DynamicConfig is a service that enables updating metric schedules, trace
+// parameters, and other configurations on an agent/SDK without having to
+// restart the instrumented application.
+service DynamicConfig {
+  rpc GetConfig (ConfigRequest) returns (ConfigResponse);
+}
+
+message ConfigRequest{
+  // Describe the version of the CS protocol to use. Required.
+  int32 version = 1;
+
+  // The resource whose config we are interested in.
+  opentelemetry.proto.resource.v1.Resource resource = 2;
+
+  // Optional. Identifies the last configuration version number
+  int32 last_config_fp = 3;
+
+  // An opaque string subject to client’s interpretation.
+  // Can includes encoded metadata such as replication level,
+  // traffic class, target information
+  string metadata = 4;
+}
+
+message ConfigResponse {
+  // Describe the version of the CS protocol to use. Should match the
+  // request’s version.
+  int32 version = 1;
+
+  // Tell the last configuration’s fingerprint. Required.
+  int32 config_fp = 2;
+
+  // Configs specific to metrics. Configs specific to traces should
+  // be a separate sub-message type.
+  message MetricConfig {
+
+    // Schedule is used to apply a particular scheduling configuration to
+    // a metric
+    message Schedule {
+      // this schedule is applied to metric names that follow this pattern
+      string metric_pattern = 1;  // Regexp
+
+      // Collection period (in seconds). Should be one of the following:
+      // * 30 seconds
+      // * 1 minute
+      // * 5 minutes
+      // * 10 minutes
+      // * 30 minutes
+      // * 1 hour
+      // * 4 hours
+      // * 12 hours
+      // * 1 day
+      // etc.
+      // Please ensure that all longer periods are divisible by all shorter
+      // ones.
+      int32 period = 2;
+
+      // Additional metadata associated with the schedule
+      string properties = 3;
+    }
+
+    repeated Schedule collecting_schedules = 1;
```

Comment:
Order is important. A metric is matched against the schedule in order and stop following patterns once the first match is found.

Reply:
done

Reply:
I think whether you can only match one schedule is an interesting question, considering we allow different metadatas associated with each Schedule. If we later allow passing back optional configuration metadata in an [ExportMetricsServiceRequest](https://github.com/open-telemetry/opentelemetry-proto/blob/master/opentelemetry/proto/collector/metrics/v1/metrics_service.proto#L35-L42), this would enable demuxing at the collector and exporter (eg, if you want to write the same metric from the same process, but at different schedules so that you can write it to multiple backends or to different monitored resource types or on higher priority networking every 10 minutes but lower priority networking only every 1 minute etc). But I'm willing to leave this an open question for now - we can always come back to it later.

Reply:
Cool idea! After talking with Ming yesterday afternoon, we've decided to leave this one on the back-burner for now and maybe return to it later? We'll focus on a simple prototype first, but this sounds like a neat feature to add in the future

Reply:
Please add documentation about cases of multiple matched schedules.

1. High-priority traffic-class 10 min schedule coexists with low-priority 1min schedule.

2. If a metric matches the same traffic class but with different period, using the shortest period.

Reply:
Documentation added